sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala

```diff
@@ -68,10 +72,57 @@ object InferWindowGroupLimit extends Rule[LogicalPlan] with PredicateHelper {
     case _ => false
   }
 
+  private def limitSupport(limit: Int, window: Window): Boolean =
+    limit <= conf.windowGroupLimitThreshold && window.child.maxRows.forall(_ > limit) &&
+      !window.child.isInstanceOf[WindowGroupLimit] &&
+      window.windowExpressions.forall(isExpandingWindow) &&
+      window.orderSpec.exists(!_.foldable) &&
+      // LimitPushDownThroughWindow have better performance than WindowGroupLimit if the
+      // window function is Rank, DenseRank and RowNumber, and Window partitionSpec is empty.
+      !(window.partitionSpec.isEmpty && supportsPushdownThroughWindow(window.windowExpressions) &&
+        limit < conf.topKSortFallbackThreshold)
+
+  private def supportsPushdownThroughWindow(
```

Comment:
We can change the method of LimitPushDownThroughWindow to `private[catalyst]` and avoid copy it.